Move FFmpeg 'run_test' configuration from platform to components

jjlawren opened the pull request:
## Description:


**Related issue (if applicable):** fixes #17495 

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7064

## Example entry for `configuration.yaml` (if applicable):
```yaml
camera:
  - platform: ffmpeg
    name: My Camera
    input: -rtsp_transport tcp -i rtsp://10.2.3.4:8554/unicast
    run_test: false

binary_sensor:
  - platform: ffmpeg_noise
    name: My Camera Sound
    input: rtsp://10.2.3.4:8554/unicast
    extra_arguments: -filter:a highpass=f=300,lowpass=f=2500,volume=volume=2 -codec:a libmp3lame -ar 16000
    initial_state: false
    duration: 2
    reset: 60
    peak: -32
    run_test: false
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

jjlawren committed:
Add 'run_test' configs to skip ffmpeg input test on startup

pvizeli commented on homeassistant/components/binary_sensor/ffmpeg_motion.py:
```diff
@@ -25,9 +25,11 @@
 CONF_CHANGES = 'changes'
 CONF_REPEAT = 'repeat'
 CONF_REPEAT_TIME = 'repeat_time'
+CONF_RUN_TEST = 'run_test'
```
Please add this to component and import this only on platforms

pvizeli commented on homeassistant/components/camera/ffmpeg.py:
```diff
@@ -22,18 +22,25 @@
-    if not await hass.data[DATA_FFMPEG].async_run_test(config.get(CONF_INPUT)):
-        return
+    if config.get(CONF_RUN_TEST):
+        if not await hass.data[DATA_FFMPEG].async_run_test(config.get(CONF_INPUT)):
```
You can short it with `manager = hass.data[DATA_FFMPEG]` like other platforms

jjlawren committed:
Shorten line which exceeds limit

jjlawren committed:
Import  key name from ffmpeg component

pvizeli commented:
I'm currently not sure if the current handling not better and we need to update the documentation. The Idea was that you can add ffmpeg sensors/camera and it will test that on startup, that makes the startup slower but for "normal" users it make the config save. If the user is sure that he doesn't need this, he can disable it on component global config with:
```yaml
ffmpeg:
  run_test: false
```

Now, both work and not. If you disable it global, the platform settings don't work. That makes it a bit weird. Why are the old settings not working for you?

jjlawren commented:
My specific use case is that I have multiple ffmpeg-based entities, where some inputs are always available and others which turn on/off. The ones that can turn on/off are sometimes not providing feeds when HA is started.

Perhaps the docs (in the components) could reflect that a global override can be set in the `ffmpeg` platform, but otherwise it will obey the per-component configs?

jjlawren committed:
Config FFmpeg tests from components instead of platform

jjlawren commented:
I've taken your suggestion and gone all-in on configuring tests on the individual components and removed the global `run_test` config at the platform level. This gives more control and I think is generally easier to understand. Updated docs to reflect this in the linked doc PR.

jjlawren commented:
@pvizeli can you comment on the latest change? I’ll need to fix the tests but want to confirm the new direction is agreeable first.

jjlawren commented:
Closing in favor of #18131.
